vapid/soa.h

```diff
@@ -196,21 +194,32 @@ namespace vapid {
             ((sort_col_by_reference(std::integral_constant<size_t, I>{})), ...);
         }
 
+        template <class C>
+        void reorder(C &c, std::vector<size_t> &o)
+        {
+            using std::swap;
+            size_t num_elems = o.size();
+            for (size_t i=0; i<num_elems; ++i)
+            {
+                while (o[i] != i)
+                {
+                    swap(c[i], c[o[i]]);
```

Comment:
This gives me an idea. We can precompute the complete sequence of swaps necessary and cache them. Then we avoid the `sort_order_tmp_ = sort_order_reference_` copy and also avoid having to do random access on this `o` vector.

Reply:
I think it would suffice to count the number of elements in each cycle of the cycle decomposition, as well as the lowest index of each cycle.

Reply:
Hum, interesting, but I suspect that the speed issue may be the non-sequential write to c (dst in your version).